util.go

```diff
@@ -144,15 +144,16 @@ func (s *segmentedBuffer) Append(input io.Reader, length uint32) error {
 
 	dst := pool.Get(int(length))
 	n, err := io.ReadFull(input, dst)
-	if err == io.EOF {
-		err = io.ErrUnexpectedEOF
+	if err != nil {
+		if err == io.EOF {
+			err = io.ErrUnexpectedEOF
+		}
+		return err
 	}
 	s.bm.Lock()
 	defer s.bm.Unlock()
-	if n > 0 {
```

Comment:
is it possible to read some bytes but also get an error? in that internally the first read made by `io.ReadFull` occurs, and partially fills the `dst` buffer, but a subsequent read fails and errors.

Is it ever useful to have the read data added to the segmented buffer when there was also an error in reading?

Reply:
We'd close the connection when this happens, so it doesn't realy matter if we correctly account for the bytes that were read.
That said, this is logic that we have in the stream, so probably we shouldn't make the `segmentedBuffer` depend on that.